_posts/2020-05-15-model-inversion-attacks/model_inversion_differential_privacy.Rmd

```diff
@@ -0,0 +1,740 @@
+    self$server_step <- function(inputs)
+      self$server_model(inputs)
+
+    function(inputs, mask = NULL) {
+      inputs %>% 
+        self$mobile_step() %>%
+        self$server_step()
+    }
+  })
+  
+}
+
+model <- target_model()
+
+```
+
+The overall model is a Keras custom model, so we train it [TensorFlow 2.x -
+style](https://tensorflow.rstudio.com/tutorials/advanced/). After ten epochs, training and validation accuracy are at
+~0.84\ and~ 0.73, respectively -- not bad at all for a 20-class discrimination task.
+
+```{r}
+loss <- loss_sparse_categorical_crossentropy
+optimizer <- optimizer_adam()
+
+train_loss <- tf$keras$metrics$Mean(name='train_loss')
+train_accuracy <-  tf$keras$metrics$SparseCategoricalAccuracy(name='train_accuracy')
+
+val_loss <- tf$keras$metrics$Mean(name='val_loss')
+val_accuracy <-  tf$keras$metrics$SparseCategoricalAccuracy(name='val_accuracy')
+
+train_step <- function(images, labels) {
+  with (tf$GradientTape() %as% tape, {
+    predictions <- model(images)
+    l <- loss(labels, predictions)
+  })
+  gradients <- tape$gradient(l, model$trainable_variables)
+  optimizer$apply_gradients(purrr::transpose(list(
+    gradients, model$trainable_variables
+  )))
+  train_loss(l)
+  train_accuracy(labels, predictions)
+}
+
+val_step <- function(images, labels) {
+  predictions <- model(images)
+  l <- loss(labels, predictions)
+  val_loss(l)
+  val_accuracy(labels, predictions)
+}
+
+
+training_loop <- tf_function(autograph(function(train_ds, val_ds) {
+  for (b1 in train_ds) {
+    train_step(b1[[1]], b1[[2]])
+  }
+  for (b2 in val_ds) {
+    val_step(b2[[1]], b2[[2]])
+  }
+  
+  tf$print("Train accuracy", train_accuracy$result(),
+           "    Validation Accuracy", val_accuracy$result())
+  
+  train_loss$reset_states()
+  train_accuracy$reset_states()
+  val_loss$reset_states()
+  val_accuracy$reset_states()
+}))
+
+
+for (epoch in 1:10) {
+  cat("Epoch: ", epoch, " -----------\n")
+  training_loop(train_ds, val_ds)  
+}
+```
+
+    Epoch:  1  -----------
+    Train accuracy 0.195090905     Validation Accuracy 0.376605511
+    Epoch:  2  -----------
+    Train accuracy 0.472272724     Validation Accuracy 0.5243119
+    ...
+    ...
+    Epoch:  9  -----------
+    Train accuracy 0.821454525     Validation Accuracy 0.720183492
+    Epoch:  10  -----------
+    Train accuracy 0.840454519     Validation Accuracy 0.726605475
+
+Now, we train the adversary.
+
+### Train adversary
+
+The adversary's general strategy will be:
+
+-   Feed its small, surrogate dataset to the on-device model. The output received can be regarded as a (highly) *compressed*
+    version of the original images.
+-   Pass that "compressed" version as input to its own model, which tries to reconstruct the original images from the sparse
+    code.
+-   Compare original images (those from the surrogate dataset) to the reconstruction pixel-wise. The goal is to minimize the
+    mean (squared, say) error.
```

Comment:
I would embolden the action sentence in each of these, e.g., **Feed its small, surrogate dataset to the on-device model.**, or something to that effect, to make the list a bit more skimmable.

Reply:
makes sense!